cluster/kubernetes/namespacer.go

```diff
@@ -0,0 +1,56 @@
+package kubernetes
+
+import (
+	"fmt"
+
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached"
+
+	kresource "github.com/weaveworks/flux/cluster/kubernetes/resource"
+)
+
+type namespaceViaDiscovery struct {
+	fallbackNamespace string
+	disco             discovery.DiscoveryInterface
+}
+
+// NewNamespacer creates an implementation of Namespacer
+func NewNamespacer(k *Kubectl, d discovery.DiscoveryInterface) (*namespaceViaDiscovery, error) {
+	fallback, err := k.getDefaultNamespace()
+	if err != nil {
+		return nil, err
+	}
+	cachedDisco := cached.NewMemCacheClient(d)
+	// in client-go v9, the call of Invalidate is necessary to force
+	// it to query for initial data; in subsequent versions, it is a
+	// no-op reset of the cache validity, so safe to call.
+	cachedDisco.Invalidate()
+	return &namespaceViaDiscovery{fallbackNamespace: fallback, disco: cachedDisco}, nil
+}
+
+func (n *namespaceViaDiscovery) correctNamespace(m kresource.KubeManifest) error {
+	namespaced, err := n.lookupNamespacedness(m)
+	switch {
+	case err != nil:
+		return err
+	case namespaced && m.GetNamespace() == "":
+		m.SetNamespace(n.fallbackNamespace)
+	case !namespaced:
+		m.SetNamespace("")
+	}
+	return nil
+}
+
+func (n *namespaceViaDiscovery) lookupNamespacedness(m kresource.KubeManifest) (bool, error) {
```

Comment:
s/`lookupNamespacedness`/`isNamespaced`/?

Reply:
I do so like stretching words out of shape .. but not super appropriate for code